###
#   File name : Combine_Seurat_and_TCR.R
#   Author    : Hyunjin Kim
#   Date      : Apr 2, 2020
#   Email     : hyunjin.kim@stjude.org
#   Purpose   : We have Seurat object from scRNA-Seq and want to combine scTCR-Seq data
#               generated by Cell Ranger VDJ to the object.
#
#   Instruction
#               1. Source("Combine_Seurat_and_TCR.R")
#               2. Run the function "combineSeuratNTCR" - specify a vector of input directories and the output file path
#               3. The combined Seurat object will be generated in the output path
#
#   Example
#               > source("The_directory_of_Combine_Seurat_and_TCR.R/Combine_Seurat_and_TCR.R")
#               > dir_list <- list.dirs("C:/Users/hkim8/SJ/SJCAR19-05/SJCAR19-05_TCR/", recursive = FALSE)
#               > combineSeuratNTCR(Seurat_RObj_path="./data/JCC212_Px5.Robj",
#                                   TCR_data_dirs=dir_list,
#                                   outFilePath="./data/JCC212_Px5_TCR_combined.Robj")
###

combineSeuratNTCR <- function(Seurat_RObj_path="./data/JCC212_Px5.Robj",
                              TCR_data_dirs,
                              outFilePath="./data/JCC212_Px5_TCR_combined.Robj") {
  
  ### load library
  if(!require(Seurat, quietly = TRUE)) {
    install.packages("Seurat")
    require(Seurat, quietly = TRUE)
  }
  
  ### load the Seurat object and save the object name
  tmp_env <- new.env()
  load(Seurat_RObj_path, tmp_env)
  obj_name <- ls(tmp_env)
  assign("Seurat_Obj", get(obj_name, envir = tmp_env))
  rm(tmp_env)
  gc()
  
  ### for each TCR data combine them to the Seurat object
  tcr <- NULL
  for(i in 1:length(TCR_data_dirs)) {
    
    ### load filtered contig annotation data
    tcr_data <- read.csv(file = paste0(TCR_data_dirs[i], "/filtered_contig_annotations.csv"),
                         stringsAsFactors = FALSE, check.names = FALSE)
    
    ### remove the -* at the end of each barcode.
    tcr_data$barcode <- sapply(strsplit(tcr_data$barcode, split = "-", fixed = TRUE), function(x) x[1])
    
    ### remove the rows that do not have CDR3 sequence
    tcr_data <- tcr_data[which(tcr_data$cdr3 != "None"),]
    
    ### remove redundant rows
    tcr_data <- tcr_data[!duplicated(tcr_data[c("barcode", "cdr3")]),]
    
    ### order by "chain" so that TRA rows come first than TRB rows
    ### and secondly, order by "CDR3 Amino Acid" sequence
    tcr_data <- tcr_data[order(as.character(tcr_data$chain), as.character(tcr_data$cdr3)),]
    
    ### annotate TRA & TRB info to the cdr3 sequences
    tcr_data$cdr3 <- paste0(tcr_data$chain, ":", tcr_data$cdr3)
    tcr_data$cdr3_nt <- paste0(tcr_data$chain, ":", tcr_data$cdr3_nt)
    
    ### now merge different TRA & TRB info to one row
    dups <- which(duplicated(tcr_data$barcode))
    if(length(dups) > 0) {
      temp <- tcr_data[dups,]
      tcr_data <- tcr_data[-dups,]
      rownames(tcr_data) <- tcr_data$barcode
      for(barcode in tcr_data$barcode) {
        idx <- which(temp$barcode == barcode)
        tcr_data[barcode,"cdr3"] <- paste(c(tcr_data[barcode,"cdr3"], temp[idx, "cdr3"]), collapse = ";")
        tcr_data[barcode,"cdr3_nt"] <- paste(c(tcr_data[barcode,"cdr3_nt"], temp[idx, "cdr3_nt"]), collapse = ";")
        tcr_data[barcode,"productive"] <- paste(c(tcr_data[barcode,"productive"], temp[idx, "productive"]), collapse = ";")
        tcr_data[barcode,"reads"] <- paste(c(tcr_data[barcode,"reads"], temp[idx, "reads"]), collapse = ";")
        tcr_data[barcode,"umis"] <- paste(c(tcr_data[barcode,"umis"], temp[idx, "umis"]), collapse = ";")
      }
    }
    
    ### only retain informative columns
    tcr_data <- tcr_data[,c("barcode", "raw_clonotype_id", "cdr3", "cdr3_nt", "reads", "umis", "productive")]
    colnames(tcr_data) <- c("barcode", "raw_clonotype_id", "cdr3_aa", "cdr3_nt", "tcr_reads", "tcr_umis", "tcr_productive")
    
    ### add time info
    temp <- strsplit(basename(TCR_data_dirs[i]), split = "_", fixed = TRUE)[[1]][-c(1,2,3)]
    if(temp[1] == "PB" || temp[1] == "BM") {
      tcr_data$time <- temp[2]
      tcr_data$type <- temp[1]
    } else if(grepl("GMP", temp[1])) {
      tcr_data$time <- "GMP"
      tcr_data$type <- "GMP"
    } else if(temp[2] == "PB" || temp[2] == "BM") {
      tcr_data$time <- temp[1]
      tcr_data$type <- temp[2]
    } else {
      tcr_data$time <- temp[1]
      tcr_data$type <- "PB"
    }
    
    ### combine the TCR data for all the data
    if(is.null(tcr)) {
      tcr <- tcr_data
    } else {
      tcr <- rbind(tcr, tcr_data)
    }
    
  }
  
  ### attach the TCR info to the metadata of the Seurat object
  obj_colnames <- colnames(Seurat_Obj@meta.data)
  tcr_colnames <- colnames(tcr)
  Seurat_Obj@meta.data <- merge(Seurat_Obj@meta.data,
                                tcr,
                                by.x = c("GexCellShort", "Time", "Type"),
                                by.y = c("barcode", "time", "type"),
                                all.x = TRUE)
  Seurat_Obj@meta.data <- Seurat_Obj@meta.data[,c(obj_colnames, setdiff(tcr_colnames, c("barcode", "time", "type")))]
  rownames(Seurat_Obj@meta.data) <- Seurat_Obj@meta.data[,"GexCellFull"]
  
  ### change the object name to the original one
  assign(obj_name, Seurat_Obj)
  rm(Seurat_Obj)
  gc()
  
  ### save the combined Seurat object
  save(list = c(obj_name), file = outFilePath)
  
}
